packages/mui-joy/src/Autocomplete/Autocomplete.tsx

```diff
@@ -0,0 +1,890 @@
+    className: classes.listbox,
+    elementType: PopperUnstyled as OverridableComponent<PopperUnstyledTypeMap<{}, 'ul'>>,
+    getSlotProps: getListboxProps,
+    externalForwardedProps: other,
+    ownerState,
+    getSlotOwnerState: (mergedProps) => ({
+      size: mergedProps.size || size,
+      variant: mergedProps.variant || 'outlined',
+      color: mergedProps.variant || 'neutral',
+    }),
+    additionalProps: {
+      anchorEl,
+      disablePortal,
+      open: popupOpen,
+      modifiers: cachedModifiers,
+    },
+    internalForwardedProps: {
+      component: AutocompleteListbox,
+    },
+  });
+
+  const [SlotLoading, loadingProps] = useSlot('loading', {
+    className: classes.loading,
+    elementType: AutocompleteLoading,
+    externalForwardedProps: other,
+    ownerState,
+  });
+
+  const [SlotNoOptions, noOptionsProps] = useSlot('noOptions', {
+    className: classes.noOptions,
+    elementType: AutocompleteNoOptions,
+    externalForwardedProps: other,
+    ownerState,
+    additionalProps: {
+      role: 'presentation',
+      onMouseDown: (event: React.MouseEvent<HTMLLIElement>) => {
+        // Prevent input blur when interacting with the "no options" content
+        event.preventDefault();
+      },
+    },
+  });
+
+  const [SlotLimitTag, limitTagProps] = useSlot('limitTag', {
+    className: classes.limitTag,
+    elementType: AutocompleteLimitTag,
+    externalForwardedProps: other,
+    ownerState,
+  });
+
+  if (limitTags > -1 && Array.isArray(startDecorator)) {
+    const more = startDecorator.length - limitTags;
+    if (!focused && more > 0) {
+      startDecorator = startDecorator.splice(0, limitTags);
+      startDecorator.push(
+        <SlotLimitTag key={startDecorator.length} {...limitTagProps}>
+          {getLimitTagsText(more)}
+        </SlotLimitTag>,
+      );
+    }
+  }
+
+  const [SlotOption, baseOptionProps] = useSlot('option', {
+    className: classes.option,
+    elementType: AutocompleteOption,
+    externalForwardedProps: other,
+    ownerState,
+    getSlotOwnerState: (mergedProps) => ({
+      variant: mergedProps.variant || 'plain',
+      color: mergedProps.color || 'neutral',
+    }),
+    additionalProps: {
+      as: 'li',
+    },
+  });
+
+  const defaultRenderOption = (optionProps: any, option: unknown) => (
+    <SlotOption {...optionProps}>{getOptionLabel(option)}</SlotOption>
+  );
+
+  const renderOption = renderOptionProp || defaultRenderOption;
+
+  const renderListOption = (option: unknown, index: number) => {
+    const optionProps = getOptionProps({ option, index });
+
+    return renderOption({ ...baseOptionProps, ...optionProps }, option, {
+      // `aria-selected` prop will always by boolean, see useAutocomplete hook.
+      selected: !!optionProps['aria-selected'],
+      inputValue,
+      ownerState,
+    });
+  };
+
+  return (
+    <React.Fragment>
+      <SlotRoot {...rootProps}>
+        {renderInput({
```

Comment:
Should we have a default implementation for `renderInput`, maybe using Joy's `Input` element?

Reply:
I did at first but I remove it because Material UI marked this prop as `required`. I also want to make it optional, what's your preference?

Reply:
Without going too much into the history of it, I would expect to be able to render the component without specifying the prop.

Reply:
Okay, now I remember another reason why `renderInput` is required. It is better to let developers define the props like `placeholder` on the Input directly, otherwise, we will have to add more props to the Autocomplete which creates complications when I want to customize the render input function.

```js
// if `renderInput` is not required, how to let developers specify placeholder props
<Autocomplete
  placeholder="..." // add placeholder prop?
/>

// what if I have a custom renderInput function like this:
<Autocomplete
  placeholder="..."
  renderInput={params => <Input placeholder="..." {...params} />
/>

// what's about other props apart from placeholder?
```

Reply:
I don't feel like this is a very strong argument 😁 But happy to start with this and see the feedback. In my opinion we should lean to supporting the simplest API if possible (which seems possible in this case). As for higher priority, isn't it up to the developer adding the `renderInput` to decide on it? In the end you are propagating all props, they can decide how to spread. Maybe I've missed here the point.

Reply:
`renderInput` could be replaced with a slot, couldn't it? It would make the API more consistent.